### ISSUE TITLE
Python Q-Learning - Add Dropout Layout causes runtime error

### ISSUE THREAD

**Issue by 0StackOverflow0** [hughperkins/DeepCL#129]
If I add a dropout layer to the Q-Learning example in python (I'm fiddling with a larger grid and network). I get the following error.

> Traceback (most recent call last):
  File "C:\Program Files\Anaconda3\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "C:\Program Files\Anaconda3\lib\threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "QLearning.pyx", line 10, in PyDeepCL.QLearner._run (PyDeepCL.cpp:17740)
RuntimeError: need to copyToDevice() before calling kernel->input

It's being placed after the first fully connected layer and before the subsequent activation function.

I'm not so familiar with OpenCL (though that's why I came to DeepCL). I'll read through the related code segments to get a better understanding of DeepCL. I can't promise that I can solve this, but I'll try to help.

I'm hoping to later create a much more complex DQN I'm interested in seeing how this regularization technique helps.

**Comment by hughperkins** [comment]
> I'm not so familiar with OpenCL (though that's why I came to DeepCL). I'll read through the related code segments to get a better understanding of DeepCL. I can't promise that I can solve this, but I'll try to help.

Ok.  The error message is something like:
- there is a CLFloatWrapper object, which is a wrapper around an underlying `cl_mem` object: https://github.com/hughperkins/EasyCL/blob/master/CLFloatWrapper.h
- the corresponding `cl_mem` object has not been created yet:
    - here's a method that creates the `cl_mem` object, for the GPU, without copying from main memory: https://github.com/hughperkins/EasyCL/blob/master/CLWrapper.cpp#L58
    - here's a method that copies from main memory buffer to the GPU `cl_mem` bufffer: https://github.com/hughperkins/EasyCL/blob/master/CLWrapper.cpp#L87
- this buffer is being passed into an OpenCL kernel, via the wrapper class CLKernel, https://github.com/hughperkins/EasyCL/blob/master/CLKernel.cpp
- here is the check and error message:
    - https://github.com/hughperkins/EasyCL/blob/master/CLKernel.cpp#L111

**Comment by 0StackOverflow0** [comment]
It's definitely happening during backprop, but I still can't quite figure out why.

I have a feeling it has to do with the limited set of experiences in the beginning,

I did notice that while `gradInputWrapper` and `outputWrapper` use `createOnDevice()` in:
https://github.com/hughperkins/DeepCL/blob/master/src/dropout/DropoutLayer.cpp#L86
`weightsWrapper` does not.

However, in `DropoutLayer::backward`, `copyToDevice()` is called, which should satisfy all input requirements because that function will `createOnDevice()` if it isn't already `onDevice`.

I don't really see anything different from conv/backward, or activate/backward.

Best I can think to do with my limited troubleshooting knowledge would be to attempt some logging and recompile.

**Comment by 0StackOverflow0** [comment]
Also, this may not be the best place to ask, but if I wanted to create a QNetwork that would take complex actions (ones with variability), would I really need to create an action set to combine all possibles action states (i.e. hundreds)?

Do you think it would be possible to have a set of outputs where a set is treated as action states, and the others as action variation (i.e. 6 outputs, [stop, left, up, right, down, distance]). I could try out customizing Qlearner and modifying the example scenario to test this out. What do you think?

**Comment by 0StackOverflow0** [comment]
All I can say further for now is that it is during the second step of Qlearner.run(), when it is training on the first experience.

**Comment by 0StackOverflow0** [comment]
I figured out a hack

If I `net.setTraining(True)`, then it works just fine.

https://github.com/hughperkins/DeepCL/blob/master/src/qlearning/QLearner.cpp#L99
I believe there should be a flag change before and after this line.

https://github.com/hughperkins/DeepCL/blob/master/src/dropout/DropoutLayer.cpp#L208
Here is the culprit I believe, but I can't yet confirm.

**Comment by hughperkins** [comment]
Ok. Thinking about it, without looking at the code, the dropout layer needs to keep the dropout tensor around, for backprop. If turning training off fixes a problem during backprop, it sounds like somehow that tensor is being recreated somehow?  You might want to check carefully then what `Wrapper` object holds the dropout tensor, ie the 1s and 0s choosing which inputs to forward to the output, and what happens to it, its lifecycle etc.

Your analysis sounds excellent! Very much appreciate your looking into this :)

**Comment by 0StackOverflow0** [comment]
From what I see, masks is declared as a `new unsigned char[]`. This is the wrapped variable.

However, it's innards are not given assignment until `generateMasks()` is called, which only happens during foward prop when `training` is set to `true`.

This is probably the issue.


Also,
From my understanding, and I may be wrong, but `training` shouldn't be set during predictions because you want the full network. However, in your alternate 'forward' path (not training), you multiple the `dropRatio` by the inputs (depressing every node), but I believe it is supposed to be 1 to 1 copy.

I was having issues compiling, not unexpected considering I am attempting to use newer version of VC. I'll try switching dev environments to implement my suggestion. However I'm new to submitting proposals, so I'll have to figure that out too.

**Comment by 0StackOverflow0** [comment]
My reruns failed >.>
My dropout got twice as far as the non, but the weights vanished (After 150 games)

With some modification testing I can say that it doesn't crash.

It's up to you on the non-training forward implementation.

However, my thought of only having the dropout during the actual training step, rather than across the whole `learnFromPast()` seems to not work.

I get vanishing weights fairly early... (Python reports `nan` for every weight) (I've been outputting them to file now to review dispersion and to see if a partially trained network after losing it's experiences can build better ones.)

I'm testing flagging training during the whole `learnFromPast()` but allowing the action prediction to be full network.

This seems to  be working as intended, but I'll wait til tomorrow to confirm.

**Comment by 0StackOverflow0** [comment]
I think there might be some conceptual issues with the experience replay.

I'll try digging into the white papers to see if I'm right. If so, I'll code up a proposal.

**Comment by 0StackOverflow0** [comment]
Sorry for making this a microblog.

I found given an environment so large, a distance reward was helpful.

I attempted to reload a network (wipe all experience) after having completed 50 and 500 rounds.

Each of these had two reloads, one with the full network and one with a perceptron (static conv network) and the loaded FC.

I found that I kept getting trapped in local minimum (a very common one)(entire q field homogenous), until I changed the last activations to sigmoid/softmax.

It's not quite there yet, but the perceptron (trained 500 rounds) and changed final activations. Seems to behave about as well as the first successful one after it had about 1200+ rounds (though this had no distance reward, and was incredibly lucky it converged; after 1800+ rounds no less)
